content/blog/python3-utf8-locale.md

```diff
@@ -0,0 +1,381 @@
+
+I was all happy getting results when then I got this wonderful surprise:
+
+```
+Task exception was never retrieved
+future: <Task finished coro=<SpeechClient.process_response() done, defined at /app/speech.py:222> exception=UnicodeEncodeError('ascii', 'X-RequestId:339e872cfcc64f6db6726fa1c7ba713c\r\nContent-Type:application/json; charset=utf-8\r\nPath:speech.hypothesis\r\n\r\n{"Text":"pre\xe7o","Offset":3900000,"Duration":6200000}', 130, 131, 'ordinal not
+in range(128)')>
+Traceback (most recent call last):
+  File "/app/speech.py", line 245, in process_response
+    response_dict = parse_body_json(response)
+  File "/app/speech.py", line 363, in parse_body_json
+    print(response)
+UnicodeEncodeError: 'ascii' codec can't encode character '\xe7' in position 130: ordinal not in range(128)
+```
+
+I was working on the whole thing for some weeks already, I was almost done and now
+all of sudden I was all like: **GODDAMMIT, STRING ENCODING PROBLEMS, MY LIFE IS OVER !!** .
+
+I worked for years with Python 2 and my experience with this is that string encoding
+problems have a great potential to make you waste a lot of good time (is the source code
+that is not utf-8 ? is the string contents ? both ? is it a string or a unicode object ?).
+
+With Python 3 I had the impression that it was all utf-8 by default (also strings are no
+more just array of bytes, they are [always unicode](https://docs.python.org/3/howto/unicode.html#the-string-type)).
+With that I imagined that the **print** function would also assume a default of being **utf-8**.
+
+Another important information is that this was happening inside a docker container using
+the [ubuntu 18.04](https://hub.docker.com/_/ubuntu/) image, the main reason for using containers
+was the fact that my code had a ton of dependencies including bindings to C++ (besides the obvious reason
+that if you are not using containers you are not cool).
+
+So giving the dependencies and the environment there was a good chance that the
+problem could be pretty complex. As I descended on a spiral of dispair I
+had a moment of clarity, these rare moments of clarity comes with years of wasting time chasing
+ghosts, so at least this time I did the basic thing of testing a very simple code
+sample before ghostbusting all over the place:
+
+```
+print("preço")
+```
+
+Yeah that is the code of the test. I was COMPLETELY sure it would work but I would test it
+anyway. As it is usual in my life, I was wrong (this happens a lot with the TDD practice
+of writing the test first and being sure that it will fail, sometimes it doesn't =P).
+
+Actually, thanks to Python I was half right =P. A quick test is my host worked, but the code
+itself was being run using docker container.
+
+So the next logic step was to try the very idiotic code inside the container, I got this:
+
+```
+>>> print("preço")
+  File "<stdin>", line 0
+
+    ^
+SyntaxError: 'ascii' codec can't decode byte 0xc3 in position 10: ordinal not in range(128)
```

Comment:
Did you saw it was a *SyntaxError* ? Maybe comparing *this piece of code* with Go's print is not fair.
I believe the problem here was that OS's locale was ASCII then Python uses that info to parse the source code.

But your original python stack trace does seem to be related to print not being able to encode a UTF-8 string. =)

Reply:
From what I read on some issues it is not the encoding of the source. The encoding of the source is the thing that by default is utf-8 in Python 3 and even forcing it to utf-8 did not solve the problem. It seems to be related to what they call "filesystem encoding" which is more of an operational system encoding, which defines the encoding of strings like filepaths that are passed on system calls etc, at least it is what I understood (and checking this filesystem encoding showed that it was actually ascii by default if no locale is set)